### DISCUSSION TITLE
Onboard configuration json is not valid

### DISCUSSION THREAD

**Discussion by JvandeKraats** [renovatebot/renovate#24647] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

36.44.1

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

I expected that the provided onboarding json configuration is valid JSON syntax.
So I expect Renovate to create a merge request in which a renovate.json file is added to the root of the repo with the given JSON as content.

### What you tried so far.

First encountered the problem in my CI/CD pipeline.
Then I tried to reproduce it locally on my machine using the Renovate CLI (36.97.1) with log level set to debug.
In both cases I got an error that the JSON provided wasn't valid.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: Using RE2 as regex engine
DEBUG: Parsing configs
DEBUG: Checking for config file in D:\Code/config.js
DEBUG: No config file found on disk - skipping
FATAL: Fatal error: Invalid JSON value: '{$schema: https://docs.renovatebot.com/renovate-schema.json, extends: [local>azure-ccoe/renovate-scheduled-pipeline#0.0.3]}'
       "err": {
         "message": "Invalid JSON value: '{$schema: https://docs.renovatebot.com/renovate-schema.json, extends: [local>azure-ccoe/renovate-scheduled-pipeline#0.0.3]}'",
         "stack": "Error: Invalid JSON value: '{$schema: https://docs.renovatebot.com/renovate-schema.json, extends: [local>azure-ccoe/renovate-scheduled-pipeline#0.0.3]}'\n    at Option.object [as parseArg] (C:\\Users\\j_v_d\\AppData\\Roaming\\npm\\node_modules\\renovate\\lib\\workers\\global\\config\\parse\\coersions.ts:36:13)\n    at handleOptionValue (C:\\Users\\j_v_d\\AppData\\Roaming\\npm\\node_modules\\renovate\\node_modules\\commander\\lib\\command.js:539:24)\n    at Command.<anonymous> (C:\\Users\\j_v_d\\AppData\\Roaming\\npm\\node_modules\\renovate\\node_modules\\commander\\lib\\command.js:566:7)\n    at Command.emit (node:events:514:28)\n    at Command.emit (node:domain:489:12)\n    at Command.parseOptions (C:\\Users\\j_v_d\\AppData\\Roaming\\npm\\node_modules\\renovate\\node_modules\\commander\\lib\\command.js:1471:18)\n    at Command._parseCommand (C:\\Users\\j_v_d\\AppData\\Roaming\\npm\\node_modules\\renovate\\node_modules\\commander\\lib\\command.js:1260:25)\n    at Command.parse (C:\\Users\\j_v_d\\AppData\\Roaming\\npm\\node_modules\\renovate\\node_modules\\commander\\lib\\command.js:909:10)\n    at Object.getConfig (C:\\Users\\j_v_d\\AppData\\Roaming\\npm\\node_modules\\renovate\\lib\\workers\\global\\config\\parse\\cli.ts:120:6)\n    at parseConfigs (C:\\Users\\j_v_d\\AppData\\Roaming\\npm\\node_modules\\renovate\\lib\\workers\\global\\config\\parse\\index.ts:25:31)\n    at processTicksAndRejections (node:internal/process/task_queues:95:5)\n    at C:\\Users\\j_v_d\\AppData\\Roaming\\npm\\node_modules\\renovate\\lib\\workers\\global\\index.ts:125:16\n    at start (C:\\Users\\j_v_d\\AppData\\Roaming\\npm\\node_modules\\renovate\\lib\\workers\\global\\index.ts:123:5)\n    at C:\\Users\\j_v_d\\AppData\\Roaming\\npm\\node_modules\\renovate\\lib\\renovate.ts:18:22"
       }
DEBUG: Missing config
DEBUG: Renovate exiting
```

The above logs are from running tool locally using the CLI (v36.97.1) with these commands:
`renovate --autodiscover=true --onboarding=true --autodiscover-filter=azure/my-project --platform=gitlab --endpoint=https://aaa.bbb.cc/api/v4/ --onboarding-config '{"$schema": "https://docs.renovatebot.com/renovate-schema.json", "extends": ["local>azure/renovate-scheduled-pipeline#0.0.3"]}'`

</details>

**Comment by rarkins** [comment]
Based on the log provided, you've used JS syntax and not JSON. For example JSON would have `"extends:"` and not simply `extends:` without quotes

**Reply by JvandeKraats** [reply]
I don't understand how the string of JSON i provided: `--onboarding-config '{"$schema": "https://docs.renovatebot.com/renovate-schema.json", "extends": ["local>azure-ccoe/renovate-scheduled-pipeline#0.0.3"]}'` ends up as:
`'{$schema: https://docs.renovatebot.com/renovate-schema.json, extends: [local>azure-ccoe/renovate-scheduled-pipeline#0.0.3]}'`

**Reply by JvandeKraats** [reply]
Just figured out that I need to escape the double quotes with `\`.
My command now is: `renovate --autodiscover=true --onboarding=true --autodiscover-filter=azure-ccoe/project-vending --platform=gitlab --endpoint=https://aaa.bbb.cc/api/v4/ --onboarding-config '{\"$schema\":\"https://docs.renovatebot.com/renovate-schema.json\",\"extends\":[\"local>azure/renovate-scheduled-pipeline#0.0.3\"]}'`

**Reply by JvandeKraats** [reply, marked as answer]
Escaping the double quotes solved the problem when calling Renovate from my machine using the CLI.
However this didn't solve the issue in my CI/CD pipelines.
What fixed the CI/CD pipeline was not just escaping the double quotes, because that still resulted in Renovate telling me that the JSON was incorrect. I've tried escaping the json in multiple ways but it didn't work.
I noticed in the output when running my pipeline that the `$schema` was being stripped away from the json, which would result in just 2 double quotes being left behind. I'm not completely sure what process is removing the `$schema` but for now just removing the whole first line: `\"$schema\":\"https://docs.renovatebot.com/renovate-schema.json\",` seemed to have solved my problem.  So I ended up with: `--onboarding-config={\"extends\":[\"local>azure-ccoe/renovate-scheduled-pipeline#0.0.3\"]}`  

I would have liked to still include the schema but for now I'm not going to spend anymore time in investigating this.
